docs/proposals/Periodic-Retrieval-of-Certificates-and-Keys-from-AKV.md

```diff
@@ -0,0 +1,121 @@
+
+### Key disabling
+
+A malicious user compromised the private key used for signing images. Bob promptly rotated the key, obtaining a new version, and disabled the compromised version. Bob configures Ratify to periodically retrieve all available versions of the key from AKV excluding any disabled versions, as a result, images signed with the disabled version fail signature verification. The build pipeline was triggered to build and sign all the images with the new version of key. As Ratify periodically retrieve all available versions of the key from AKV, it ensures successful verification of images signed with the latest key version without requiring manual configuration adjustments. By following these actions, Bob maintains security while seamlessly transitioning to the new version of key.
+
+### Immediate retrieval of certificates and keys 
+
+In urgent situations, such as when the private key is compromised, both Alice and Bob prefer not to wait for the next certificate and key retrieval cycle. They expect that images signed with compromised keys will fail validation immediately to avoid potential security attacks, and images signed with the rotated version can be validated successfully to prevent potential service downtime. To achieve this, they want to trigger Ratify to retrieve certificates and keys promptly, ensuring that the necessary updates are applied. This way, images can be verified with the correct configurations.
+
+## User experiences
+
+Typically, users do not need to manually update the configuration unless they specifically want to trigger an immediate retrieval. Ratify periodically retrieves the certificates and keys from Azure Key Vault (AKV), ensuring that it uses the most up-to-date available versions. Additionally, Ratify filters out any disabled versions of certificates and keys during this process.
+
+### Periodic retrieval of certificates and keys from AKV
+
+To periodically retrieve certificates and keys from AKV, users configure the `KeyManagementProvider` without specifying the version for certificates and keys. Users should have the ability to customize the time interval at which the retrieval process occurs, allowing them to override the default interval (e.g., 24 hours). As part of this enhancement, a new parameter named `retrievalInterval` will be introduced for AKV as the key management provider.
```

Comment:
Are we sure we shouldn't just introduce a new `periodicRetreval` flag? There might be a scenario where user pins to a specific version but they still want periodic retrieval to check if that version is in disabled state.

Reply:
> Are we sure we shouldn't just introduce a new `periodicRetreval` flag? There might be a scenario where user pins to a specific version but they still want periodic retrieval to check if that version is in disabled state.

If users specify a version, then it will be periodic retrieval of the latest status. if users do not specify a version, then it means periodic retrieval all available versions excluding the disabled versions.

Reply:
Here is the proposed behavior in the updated PR
- If users specify a version, then sync the state of this specific version will be done periodically without need of any parameters.
- if users specify only name, then by default the latest version and one previous version will be synced.
   - users can set the parameter `previousVersionCount` to 0 to only sync the latest version
   - users can set the parameter `previousVersionCount` to other values to sync as many as previous versions they want.

I am open to other parameter names or other parameters that can be used for achieving the same goal. @akashsinghal @susanshi @duffney

Reply:
What is the purpose of the n and n-1 being loaded for latest by default? Is this primarily for keyed scenarios where if a new version is generated, the previous version will continue to be used for verification considering the new key may not be used for signing yet?

Reply:
@akashsinghal Yes, you are right. The previous version may be used for verification as not all images are signed using the latest version right after rotation, so keeping two versions will ensure a smooth transition to the latest version. This is primarily for Cosign keyed scenarios.